bundle/src/main/java/com/adobe/acs/commons/wcm/vanity/impl/VanityURLServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+
+	@Reference
+	QueryBuilder queryBuilder;
+
+	public boolean dispatch(SlingHttpServletRequest request, SlingHttpServletResponse response) throws ServletException, IOException, RepositoryException {
+		if (request.getAttribute(VANITY_DISPATCH_CHECK_ATTR) != null) {
+			log.trace("Processing a previously vanity dispatched request. Skipping...");
+			return false;
+		}
+
+		request.setAttribute(VANITY_DISPATCH_CHECK_ATTR, true);
+
+		final String requestURI = request.getRequestURI();
+		final RequestPathInfo requestPathInfo = new PathInfo(request.getResourceResolver(), requestURI);
+
+		// Manually strip off any selectors or extensions from the URL
+		final String resourcePath = StringUtils.substringBefore(requestPathInfo.getResourcePath(), ".");
+		// Map the incoming URL to remove any prefix
+		final String candidateVanity = request.getResourceResolver().map(resourcePath);
+		final String pathScope = StringUtils.removeEnd(resourcePath, candidateVanity);
+
+		log.debug("Candidate vanity URL to check and dispatch: [ {} ]", candidateVanity);
+
+		// Check if...
+		// 1) the candidateVanity and the requestURI are the same; If they are it means the request has already
+		// gone through resource resolution and failed so there is no sense in sending it through again.
+		// 2) the candidate is in at least 1 sling:vanityPath under /content
+		if (!StringUtils.equals(candidateVanity, requestURI) && isVanityPath(pathScope, candidateVanity, request)) {
+			log.debug("Forwarding request to vanity resource [ {} ]", candidateVanity);
+
+			final RequestDispatcher requestDispatcher = request.getRequestDispatcher(candidateVanity);
+			requestDispatcher.forward(new ExtensionlessRequestWrapper(request), response);
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks if the provided vanity path is sling:vanityPath under the mapped path prefix
+	 *
+	 * @param pathScope The content path to scope the vanity search too.
+	 * @param vanityPath Vanity path that needs to be validated.
+	 * @param request SlingHttpServletRequest object used for performing query/lookup
+	 * @return return true if the vanityPath is a registered sling:vanityPath under /content
+	 */
+	protected boolean isVanityPath(String pathScope, String vanityPath, SlingHttpServletRequest request) throws RepositoryException {
+		final long start = System.currentTimeMillis();
+
+		final Map<String, String> params = new HashMap<>();
+
+		// Limit to <pathScope> to get 1/2 to multi-tenant support
+		params.put("path", StringUtils.defaultIfEmpty(pathScope, DEFAULT_PATH_SCOPE));
+		params.put("property", NameConstants.PN_SLING_VANITY_PATH);
+		params.put("property.value", vanityPath);
+		params.put("p.limit", "1");
+		params.put("p.guessTotal", "true");
+
+		log.debug("Searching for vanity path [ {} ] under [ {} ]", vanityPath, pathScope);
+
+		final Query query = queryBuilder.createQuery(PredicateGroup.create(params), request.getResourceResolver().adaptTo(Session.class));
```

Comment:
> Still interested wrt to the performance profile property indexes w ~0-1 or 1 matches.

It very much depends on query. 

    select * from [nt:base] where [jcr:uuid] = 'foo'

If the index used is a unique index and its a Mongo based setup then it would result in one read calls in Mongo for "/oak:index/uuid/:index/<uuid entry>". 

    select * from [nt:base] where [sling:resourceType] = 'type1'

However if its a property index then it may result in multiple Mongo calls as the index storage mirrors the content path of indexed content. So here it would have to read a structure like

```
oak:index
  + slinResourceType
     + :index
        + type1
            + content
               + en
                  ... //node structure mirroring indexed content
```
So for reading the first result for 'type1' it would need to read all those child nodes (if they have not been found in cache).